taigapy/client_v3.py

```diff
@@ -231,6 +231,93 @@ def _ensure_dataset_version_cached(self, permaname: str, version: int) -> bool:
                     )
         return True
 
+    def _get_file_storage_type(
+        self, metadata: MinDataFileMetadata
+    ) -> TaigaStorageFormat:
+        if metadata.type == "HDF5":
+            return TaigaStorageFormat.HDF5_MATRIX
+        if metadata.type == "Columnar":
+            return TaigaStorageFormat.CSV_TABLE
+        if metadata.type == "Raw":
+            # Refactored into separate function due to needing this in 2 places. The following
+            # 2 comments were originally added by Jessica.
+            # NOTE: Upload files with HDF5_MATRIX format is stored in taiga as Raw format but its custom_metadata dict 'client_storage_format' key holds info about its format as RAW_HDF5_MATRIX (see _upload_uploaded_file()). This is confusing and it would be nice if client and api data types match. Also, it's possible for custom_metadata to somehow be None so need to account for that too...
+            # TODO: Figure out how custom_metadata can be None. My guess is uploads that used old client?
+            value = (
+                metadata.custom_metadata.get(
+                    "client_storage_format", TaigaStorageFormat.RAW_BYTES
+                )
+                if metadata.custom_metadata
+                else TaigaStorageFormat.RAW_BYTES
+            )
+            return TaigaStorageFormat(value)
+        else:
+            raise Exception(f"unknown type: {metadata.type}")
+
+    def _get_dataframe_offline(
+        self,
+        id: Optional[str],
+        name: Optional[str],
+        version: Optional[DatasetVersion],
+        file: Optional[str],
+    ):
+        print(
+            cf.orange(
+                "You are in offline mode, please be aware that you might be out of sync with the state of the dataset version (deprecation)."
+            )
+        )
+        if id is None:
+            if name is None:
+                print(cf.red("If id is not specified, name must be specified"))
+                return None
+
+            if version is None:
+                print(cf.red("Dataset version must be specified"))
+                return None
+
+            id = f"{name}.{version}/{file}"
+
+        metadata = self.datafile_metadata_cache.get(id, None)
+        parsed_datafile_id = _parse_datafile_id(id)
+
+        assert (
+            parsed_datafile_id
+        ), f"{id} doesn't look like a well qualified taiga datafile ID"
+
+        canonical_id = self.canonical_id_cache.get(id, None)
```

Comment:
This code and below looks a lot like `self._get(...)` and I think I understand why you copied it here (needed to in order to eliminate the call `self.download_to_cache()` and instead fetch the value from the cache )

At the same time, we now have two code paths depending on whether we're online or not and I worry that they might fall out of sync. And given that most of the time we'll be testing changes while online, we might notice if we make a change that breaks the offline mode. (ie: if we added another storage format)

What would you think about folding things together by adding an extra parameter to `_get()` and `download_to_cache()`? 

I'm imagining something like:

```
def def _get(self, datafile_id: str, only_use_cache=False) -> pd.DataFrame:
   ...
   path = self.download_to_cache(datafile_id, LocalFormat.HDF5_MATRIX, only_use_cache=only_use_cache)
   result = read_hdf5(path)
   ...

def download_to_cache(
        self, datafile_id: str, requested_format: Union[LocalFormat, str],
        only_use_cache = False
    ) -> str:

       canonical_id = self.get_canonical_id(datafile_id)
        key = repr((canonical_id, requested_format))
        path = self.internal_format_cache.get(key, None)
        if path:
            return path
        assert not only_use_cache, f"Expected {key} to be cached, but it was not!"
        ...
```

I think this should minimize the amount of code that is specific to when the client is offline.
     
What do you think?

Reply:
I can definitely see how this would help keep things in sync. I'll make this change.

Reply:
Thank you!